knowledge-base/fiddlercore-versus-fiddler-desktop-certificate-creation.md

```diff
@@ -0,0 +1,142 @@
+---
+title: Certificate Creation in Fiddler Core and Fiddler Desktop
+description: Comparing the Certificate Creation Methods used in Fiddler Core and Fiddler Desktop
+type: how-to
+page_title: FiddlerCore versus Fiddler Desktop Certificate Creation
+slug: fiddlercore-versus-fiddler-desktop-certificate-creation
+position: 
+tags: Certificates, Bouncy Castle, MakeCert
+ticketid: 1429862
+res_type: kb
+---
+
+## Environment
+<table>
+	<tbody>
+		<tr>
+			<td>Product Version</td>
+			<td>4.6.20191.7809</td>
+		</tr>
+		<tr>
+			<td>Product</td>
+			<td>Progress® Telerik® FiddlerCore Embedded Engine</td>
+		</tr>
+	</tbody>
+</table>
+
+
+## Description
+FiddlerCore uses different APIs for certificate creation than the Desktop Version. By default, FiddlerCore includes the CertMaker and BCMakeCert assemblies for use with the Bouncy Castle API but doesn't keep the certificate for later use.
+
+In contrast, the Default Certificate creation in the Desktop Version uses the MakeCert.exe with the Windows Crypto API and does keep the Certificate for later use.
+
+## Suggested Workarounds
+Use the 
+
+### Using MakeCert.dll with Application (Recommended)
+In this example, the MakeCert.dll is used with the ICertificateProvider Interface to store the Certificate Information. For example, in a console application prior to setting up FiddlerCore, call the `EnsureRootCertificate` method like below.
+
+``` csharp
+	private static void EnsureRootCertificate()
+	{
+		string certMakerPath = Path.Combine(assemblyDirectory, "lib", "CertMaker.dll");
+		FiddlerApplication.Prefs.SetStringPref("fiddler.certmaker.assembly", certMakerPath);
+		CertMaker.EnsureReady();
+		ICertificateProvider5 certificateProvider = (ICertificateProvider5)CertMaker.oCertProvider;
+
+		string rootCertificatePath = Path.Combine(assemblyDirectory, "..", "..", "RootCertificate.p12");
+		string rootCertificatePassword = "S0m3T0pS3cr3tP4ssw0rd";
+
+		if (File.Exists(rootCertificatePath))
+		{
+			certificateProvider.ReadRootCertificateAndPrivateKeyFromPkcs12File(rootCertificatePath, rootCertificatePassword);
+		}
+		else
+		{
+			certificateProvider.CreateRootCertificate();
+			certificateProvider.WriteRootCertificateAndPrivateKeyToPkcs12File(rootCertificatePath, rootCertificatePassword);
+		}
+
+		if (!CertMaker.rootCertIsTrusted())
+		{
+			CertMaker.trustRootCert();
+		}
+	}
+
+	private static void Main()
```

Comment:
Why is the here? To give more context? If this is the goal, I am not sure it's achieving it - I mean the method names are not so self-descriptive, especially for a newcomers (SAZ-what?).

Reply:
Yeah, I was trying to add context but am not sure it does either. I was thinking of adding the sample from the blog post which provides the full context. We are reaching out to the KB lead to find out more on where we should host the sample code for download. In this case it might be better to just link to the download on the Forum. Will update once I know more.

Reply:
Removed the `main` functon. It did seem unnecessary since we are including a Repository for Demos.

Reply:
Will just let the code in the KB speak for itself.